File: Editorial/3212_Count_Submatrices_With_Equal_Frequency_of_X_and_Y/C/1_Approach_2D_Prefix_Sum.c
int numberOfSubmatrices(char** grid, int gridSize, int* gridColSize) {
    int n = gridSize, m = gridColSize[0];
    int ans = 0;
    int sum[n + 1][m + 1][2];

    memset(sum, 0, sizeof(sum));

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (grid[i][j] == 'X') {
                sum[i + 1][j + 1][0] =
                    sum[i + 1][j][0] + sum[i][j + 1][0] - sum[i][j][0] + 1;
                sum[i + 1][j + 1][1] = 1;
            } else if (grid[i][j] == 'Y') {
                sum[i + 1][j + 1][0] =
                    sum[i + 1][j][0] + sum[i][j + 1][0] - sum[i][j][0] - 1;
                sum[i + 1][j + 1][1] = sum[i + 1][j][1] | sum[i][j + 1][1];
            } else {
                sum[i + 1][j + 1][0] =
                    sum[i + 1][j][0] + sum[i][j + 1][0] - sum[i][j][0];
                sum[i + 1][j + 1][1] = sum[i + 1][j][1] | sum[i][j + 1][1];
            }
            if (sum[i + 1][j + 1][0] == 0 && sum[i + 1][j + 1][1] == 1) {
                ans++;
            }
        }
    }
    return ans;
}

File: Editorial/3201-3300/3212_Count_Submatrices_With_Equal_Frequency_of_X_and_Y/CSharp/1_Approach_2D_Prefix_Sum.cs
public class Solution {
    public int NumberOfSubmatrices(char[][] grid) {
        int n = grid.Length, m = grid[0].Length;
        int ans = 0;
        int[,,] sum = new int[n + 1, m + 1, 2];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 'X') {
                    sum[i + 1, j + 1, 0] =
                        sum[i + 1, j, 0] + sum[i, j + 1, 0] - sum[i, j, 0] + 1;
                    sum[i + 1, j + 1, 1] = 1;
                } else if (grid[i][j] == 'Y') {
                    sum[i + 1, j + 1, 0] =
                        sum[i + 1, j, 0] + sum[i, j + 1, 0] - sum[i, j, 0] - 1;
                    sum[i + 1, j + 1, 1] =
                        (sum[i + 1, j, 1] | sum[i, j + 1, 1]) == 1 ? 1 : 0;
                } else {
                    sum[i + 1, j + 1, 0] =
                        sum[i + 1, j, 0] + sum[i, j + 1, 0] - sum[i, j, 0];
                    sum[i + 1, j + 1, 1] =
                        (sum[i + 1, j, 1] | sum[i, j + 1, 1]) == 1 ? 1 : 0;
                }

                if (sum[i + 1, j + 1, 0] == 0 && sum[i + 1, j + 1, 1] == 1) {
                    ans++;
                }
            }
        }
        return ans;
    }
}
